### ISSUE TITLE
Passkey support

### ISSUE THREAD

**Issue by ThinkChaos** [keepassium/KeePassium#297]
Keepass[XC] don't support this yet, but I wanted to open an issue here to track it anyways as progress could be made even before the others apps do, for instance by storing passkeys in custom entry fields.

As per [this blog post](https://blog.millerti.me/2023/06/10/wwdc23-passkeys/#third-party-passkey-providers), iOS and macOS have introduced a passkey provider API already:

> Apple has also publicly revealed their own passkey provider API in the form of additional functionality added to the existing [ASCredentialProviderViewController](https://developer.apple.com/documentation/authenticationservices/ascredentialproviderviewcontroller), “a view controller that a password manager app uses to extend Password AutoFill:”
> 
> - `prepareInterface(forPasskeyRegistration:)` ([link](https://developer.apple.com/documentation/authenticationservices/ascredentialproviderviewcontroller/4172626-prepareinterface))
> - `prepareInterfaceToProvideCredential(for:)` ([link](https://developer.apple.com/documentation/authenticationservices/ascredentialproviderviewcontroller/4172627-prepareinterfacetoprovidecredent))
> - `provideCredentialWithoutUserInteraction(for:)` ([link](https://developer.apple.com/documentation/authenticationservices/ascredentialproviderviewcontroller/4172628-providecredentialwithoutuserinte))
>
> Documentation for these methods are light on details so that’s about all I could make sense of. These new instance methods say they’ll become available for third-party passkey providers to use in iOS 17, iPadOS 17, and macOS 14.

**Comment by keepassium** [comment]
Yes, good point, thank you!

**Comment by hkaancaliskan** [comment]
For tracking: https://github.com/keepassxreboot/keepassxc/issues/1870

**Comment by jasperweiss** [comment]
See [draft pull request](https://github.com/keepassxreboot/keepassxc/pull/8825) for current implementation

**Comment by keepassium** [comment]
@jasperweiss , @hkaancaliskan , the nudges are well noted, thanks :)

**Comment by hkaancaliskan** [comment]
@keepassium on keepassxc side passkey support merged :) any progress on keepassium side?

**Comment by jasperweiss** [comment]
> @keepassium on keepassxc side passkey support merged :) any progress on keepassium side?

@hkaancaliskan It's not released yet. Some breaking changes might still be made so it's probably a good idea for KeePassium to wait for KeePassXC to release their final implementation.

**Comment by Calmquist** [comment]
@hkaancaliskan @jasperweiss KeePassXC's passkey release is currently pending this issue: https://github.com/keepassxreboot/keepassxc/issues/10197

**Comment by Calmquist** [comment]
It appears that KeePassXC moved that issue to 2.8.0 and released 2.7.7.

**Comment by nitn3lav** [comment]
KeePassXC passkey support is now [released](https://keepassxc.org/blog/2024-03-10-2.7.7-released/) and seems to be working great (tested with GithHub and passkey.org in Firefox)

**Comment by aliaksandrsen** [comment]
Now it would be nice to have passkey support in KeePassium

**Comment by keepassium** [comment]
Perhaps I should explain why this takes so long.

- Passkeys can be created and used only via AutoFill API. This is different from passwords, which can be created in the main app and then just used in AutoFill. So we need to have the database editable in AutoFill.
- AutoFill process has very limited memory allowance — only around 120 MB for everything: the code, libraries and the database. (This is a system-imposed constraint.)
- KeePassium core was written in early 2018, before the introduction of AutoFill in iOS 12. Memory consumption was not a concern at the time, so the app loads the database in a rather memory-hungry way. It loads the whole encrypted file, then decrypts the full thing in memory, then processes the underlying XML content using a DOM parser (which keeps the whole thing — you guessed it — in memory). 
- This works fine in the main app, but AutoFill can barely _load_  a modest 3-5 MB database before hitting the 120 MB threshold and getting terminated by the system. Generating and encrypting an updated version of the database will guarantee a crash, it just won't fit.

In order to make database editable in AutoFill, we need to rewrite database processing to use small data chunks. This should drastically reduce the memory consumption of AutoFill. However, this requires rewriting the DB processing code from the ground up — which is quite a hefty task. 

Albeit slowly, we are walking the path. A few days ago I finished optimization of the XML parsing process, it will make AutoFill a bit leaner already in the next update. But we also need to switch to chunked encryption/decryption. This is [happening](https://github.com/keepassium/StreamKit/tree/keepassium) right about now, but will take a couple of months to complete (unless something else requires urgent firefightning…) Might also have to offload entry attachments to disk temporarily, since attachments are the worst offenders memory-wise.

Once the ground work is done, it will unlock entry editing in AutoFill (#87). And then there will be passkeys.

**Comment by ThinkChaos** [comment]
Thanks for the detailed update and working on this, sounds like a lot of work!

> Might also have to offload entry attachments to disk temporarily, since attachments are the worst offenders memory-wise.

Just an idea that might make this doable with less complexity, obviously I'm not familiar with any of the details so I don't really know if it would be practical or even simpler, but here it goes:
Maybe you could save the position+length in the stream where the attachment (or whatever untouched data) is. Then when saving the modified DB, also re-create/reset the read stream and seek to get the untouched data back.  
I'm assuming it'd be simpler since it'll avoid needing code to juggle extra files, and could re-use the necessary reading code.

**Comment by keepassium** [comment]
@ThinkChaos , thanks! Yes, this is also an option. I'm a bit concerned about performance overheads, though, since skipping to a position of a multilayer (cypher + gzip + another cypher) stream implies doing all the work nevertheless. And for the earlier kdbx3 format getting to an attachment is really a piece of cake. Really thick multi-layer cake: external cypher + gzip + xml + base64 + inner cypher + gzip again :) 

I've been thinking to just save chunks of attachments when loading the database. Each chunk padded with a random-length garbage to obscure its size, and encrypted with a random key, unique for each chunk. When the time comes to reassemble and save the database, it would be just reading and decrypting the cached chunks. And if the device is compromised and someone gets to the cached chunks, these would be just piecemeal binary blobs with random names and sizes.

**Comment by Jerroder** [comment]
> Passkeys can be created and used only via AutoFill API. This is different from passwords, which can be created in the main app and then just used in AutoFill. So we need to have the database editable in AutoFill.

I don't know how much work it would be, or if it would be worth it, to at least temporarily only allow using already existing passkeys and not allow creating them?

**Comment by keepassium** [comment]
> I don't know how much work it would be, or if it would be worth it, to at least temporarily only allow using already existing passkeys and not allow creating them?

This was the case for TOTPs for a while :) But passkey implementation seems far from trivial, so I'd rather deep-dive into it only once. Otherwise there will be much time left on context switching.

**Comment by unoukujou** [comment]
> This is [happening](https://github.com/keepassium/StreamKit/tree/keepassium) right about now, but will take a couple of months to complete (unless something else requires urgent firefightning…)

Hey. Is there any ETA on this? Just wondering about the progress because I'm really looking forward to it.

**Comment by danielwagn3r** [comment]
Would be great to have passkey support in AutoFill line Apples and Microsofts Tool already offer.

![image](https://github.com/user-attachments/assets/9115189a-8cc8-48c9-88d8-315318faedc7)

**Comment by fizmhd** [comment]
Hi Team, any update ?? 
On passkey

**Comment by infirms** [comment]
Looking forward for passkeys

**Comment by keepassium** [comment]
Look what's coming in the next beta :)
Just need to make sure changes are preserved even if AutoFill crashes while saving…

https://github.com/user-attachments/assets/0f34f386-c868-4d45-8d59-f6b609a14820

**Comment by keepassium** [comment]
Just uploaded 1.54.156 to TestFlight (beta testing) and a macOS build [here](https://github.com/keepassium/KeePassium/releases/tag/1.54.156). 

It covers the main functionality: creating passkeys and using them to sign in. Some pieces are still work-in-progress, though. Specifically, recovering when AutoFill runs out of memory and crashes while saving. (This won't corrupt the DB, just no passkey would not be created.)

Early feedback is welcome!

**Comment by jasperweiss** [comment]
Great to see this in KeePassium! I have just downloaded the TestFlight build. I found a couple of issues:
- The "Create a Passkey?" prompt seems to be covered by the keyboard ~on newly created databases~ sometimes (see `debug.mp4`). It seems like the passkey prompt is supposed to dismiss the keyboard/search but sometimes the keyboard opens a little later than the passkey prompt. So they appear to be racing each other.
- At the bottom of the entity list it says `conversation entity: ?`
- When signing in with a passkey, the autofill dialog filters out entities that don't contain a passkey but if there are no entities with a passkey at all in the selected database, all entities are shown as an option and tapping on them exits the autofill dialog without doing anything (see `debug2.mp4)`
- Perhaps not an issue per se, but it would be nice if you could select the folder and edit other attributes of the entity before saving (as if creating a new entity within the app itself) or you end up with a root folder full of iconless entities with the domain as their name (as is the case with KeePassXC unfortunately)

https://github.com/user-attachments/assets/1da7ed05-35b1-453c-8559-38103e0f49ef

https://github.com/user-attachments/assets/c7070497-1a7d-4f91-bc4d-9c1e52778a82

**Comment by keepassium** [comment]
@jasperweiss , thank you! I think I've fixed most of these for the upcoming beta.

> At the bottom of the entity list it says `conversation entity: ?`

In English it says `Caller ID: ?` (that is, the identify of the page or app that asked for auto-filling). I have fixed the question mark, but is "Gespreksidentiteit" as suitable translation?

> Perhaps not an issue per se, but it would be nice if you could select the folder and edit other attributes of the entity before saving (as if creating a new entity within the app itself) or you end up with a root folder full of iconless entities with the domain as their name (as is the case with KeePassXC unfortunately)

Yes, the current thing is really minimalistic. Unfortunately, this is a balancing act. Too many nice-to-haves and we end up with a full-blown entry editor, when most people might only need a "Save" button. So this will evolve based on feedback, I just want to see which parameters really _need_ to be customizable immediately in AutoFill, and which can be left for later. (For instance, favicon downloading is likely in the latter camp.)

**Comment by jasperweiss** [comment]
> In English it says Caller ID: ? (that is, the identify of the page or app that asked for auto-filling). I have fixed the question mark, but is "Gespreksidentiteit" as suitable translation?

It seems a bit out of place. I would read that as my identity for a given conversation. The other translations seem to translate "Caller ID" quite literally. The German and French translations are "Anruf-ID" and "ID D'appel" respectively, which you'd really use in the context of a telephone conversation, I think. The Spanish translation "Identificación de la persona que llama" quite literally means "identification of the person who is calling".

In the context of "Relying party" (which I think is the phrasing used in the spec), "Vertrouwende partij" seems appropriate in Dutch. "Vertrauende Partei" in German or "Partie de confiance" in French. These all roughly mean "Trusting party" but I can only really comment on the Dutch translation.

**Comment by Jerroder** [comment]
With the latest build (1.54 (157)) I tried creating a Passkey for Paypal but the button to add it to an existing entry was greyed out
![IMG_1174](https://github.com/user-attachments/assets/20dc5bdb-3cd5-4e39-8134-2b2c378cc05e)

I created a new entry, but couldn't save the updated database
![IMG_1176](https://github.com/user-attachments/assets/e3b9382a-15c9-417d-9eba-478f7e50a8a4)

so I just clicked "saved as..." in order to create a new one, it did and came back to this screen and I couldn't do anything. The new entry is in the new database, but when I go back to Paypal after this, it said it couldn't create a Passkey.
![IMG_1177](https://github.com/user-attachments/assets/3a046a64-20f2-4695-a37c-69b0563e1909)

just to try, I tried to delete the newly created database, but couldn't
![IMG_1178](https://github.com/user-attachments/assets/6aa8f6ec-7261-4a52-8584-d2e2048fbc8a)

So the quick fill has permissions to create a database, but not to modify or delete an existing one?

**Comment by keepassium** [comment]
> It seems creating a passkey using Chrome on a desktop (MacBook) by scanning the QR code does not work.

@jasperweiss , thanks, I never thought of testing this and indeed it does not work via QR codes.After an hour of testing, I did not find any obvious problems, so this will need a deeper dive. (Given the season, probably already in January.)

**Comment by Jerroder** [comment]
> Just https://github.com/keepassium/KeePassium/issues/297#issuecomment-2515094106 how many people need that function (strongly enough to ask about it).

It is a nice-to-have feature indeed but not must-have for me. Depending on the entry I'll probably need to go in there manually to rename or remove some older TOTP methods or sormthing.

> AutoFill responded to system with an "existing passkey selected" mode. But the system expected a "passkey created" kind of response. (They have different formats.)
So PayPal rightfully said that passkey creation failed.

That's what I thought, it makes sense.

> Is the original database stored in "On My iPhone / KeePassium"? Was the new one saved there as well?

Yes and yes.

**Comment by keepassium** [comment]
>> Is the original database stored in "On My iPhone / KeePassium"? Was the new one saved there as well?
>
> Yes and yes.

Good! Then it looks like bad luck mixed with a bug :)

Due to [some technical quirks](https://github.com/keepassium/KeePassium/issues/381), AutoFill cannot access KeePassium's folder easily. If you restart the device and open AutoFill before launching the main KeePassium app, AutoFill would get a "permission denied" from the system. As a plan-B, AutoFill loads an internal backup file instead — in read-only mode. (It should also show a message that the database is unreachable.)

And then there is a bug. Since AutoFill is very new to this whole "database editing" thing, it does not check whether the file is read-only — it just tries to write it. And fails, because the database is actually unreachable.

Long story short, there should have been a "Database is read only" message instead of passkey creation dialog. We'll get that fixed. Thanks again!

**Comment by Sereby88** [comment]
@keepassium 
„Is the original database stored in "On My iPhone / KeePassium"? “ 
I use it from a WebDAV share and read / write to it from multiple devices. Ist this planned?

and yes, adding passkeys to an existing entry would be my wish, too

**Comment by keepassium** [comment]
@Sereby88 , noted. Regarding WebDAV: [already done](https://support.keepassium.com/kb/sync-webdav/).

**Comment by Sereby88** [comment]
@keepassium 
The WebDAV part was a linked question to passkeys support. You wrote about local files and problems that may occur if not locally stored. 

I would use the database from a WebDAV share (like now) and would like to use Passkeys and add them to existing entries to the database on the share. Hopefully this will work without having to change to local database per device

I hope you understand what I mean :-)

**Comment by keepassium** [comment]
@Sereby88 , the problem was because of a locally stored DB :) Remote files should work even better in this regard.

**Comment by keepassium** [comment]
@Jerroder , I was wrong, failure to save was a different issue (now filed as #406). Moving the database out of `On My iPhone / KeePassium` to anywhere else should fix saving.

**Comment by danielwagn3r** [comment]
First of all great you've finally made initial passkey support available 👏

I've played around a bit with it the last days. And I've seen some scenarios where passkey generation didn't complete successfully. An public reproducable sample is adding a passkey to an Auth0 account:

Step 1 (browser):
![image](https://github.com/user-attachments/assets/c3486fb4-767d-4700-84f1-d39114c9b8dd)

Step 2 (iphone):
Scan of QR-Code
Jump into Keepassium, save passkey to database (stored in onedrive)
passkey doesn't seem to be saved - can't find it afterwards in the database

Step 3 (browser):
![image](https://github.com/user-attachments/assets/95c7b830-401d-4572-baf5-83dbb596282a)

**Comment by keepassium** [comment]
@danielwagn3r , thank you for the feedback. Yes, passkey registration via QR codes is broken at the moment. I have registered it as #408 so it does not get lost in this thread. (Chances are, there will probably be many more passkey-related issues :)

**Comment by acuteaangle** [comment]
Hello!
With [Strongbox no longer an option for me][1], is there anything I can do help bring this functionality to Keepassium?

[1]: https://old.reddit.com/r/strongbox/comments/1jaljzn/strongbox_was_taken_over_by_the_company_applause/

**Comment by KucharczykL** [comment]
> Hello! With [Strongbox no longer an option for me](https://old.reddit.com/r/strongbox/comments/1jaljzn/strongbox_was_taken_over_by_the_company_applause/), is there anything I can do help bring this functionality to Keepassium?
> 
> EDIT: I’ve [been informed](https://old.reddit.com/r/strongbox/comments/1jaljzn/strongbox_was_taken_over_by_the_company_applause/mhxa9fp/) that Keepassium _does_ have Passkey support; however, I am still unsure if this applies to iOS, as I have been unable to find my existing KeePassXC / Strongbox passkeys in the interface, nor any other reference to passkeys.

Passkeys work on iOS with KeePassium, just used it to log in to GitHub.

**Comment by acuteaangle** [comment]
Whoops, you are correct. I don’t know how I missed it before.

Looks like I’m switching to Keepassium then :D

**Comment by ThinkChaos** [comment]
Closing this as the feature  was released in [1.54.156](https://github.com/keepassium/KeePassium/releases/tag/1.54.156).  
Thank you for your work @keepassium!
